### ISSUE TITLE
should_implement_trait triggers on unrelated clone() -> Result<T>

### ISSUE THREAD

**Issue by pronebird** [rust-lang/rust-clippy#14562]
### Summary

The following warning is counter-productive and should be fixed to understand that if return type is not `Self` then it should not be emitted.

Consider this, I have my own fallible `fn clone() -> Result<Self>`. This is *the best name* for this method. It must not be anything else.

```
warning: method `clone` can be confused for the standard trait method `std::clone::Clone::clone`
   --> src\security\sid.rs:247:5
    |
247 | /     pub fn clone(&self) -> Result<Self> {
248 | |         unsafe { Self::copy_from(self.inner) }
249 | |     }
    | |_____^
    |
```

### Lint Name

should_implement_trait

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

### Version

```text

```

**Comment by samueltardieu** [comment]
Out of curiosity, did you consider `try_clone()`? This would reflect the fact that the cloning operation can fail, and would be more in line with the standard library naming conventions (`try_from()`, `try_into()`, …).
